documentation-support/src/main/java/tech/picnic/errorprone/documentation/BugPatternTestExtractor.java

```diff
@@ -0,0 +1,227 @@
+package tech.picnic.errorprone.documentation;
+
+import static com.google.errorprone.matchers.Matchers.hasAnnotation;
+import static com.google.errorprone.matchers.Matchers.instanceMethod;
+import static com.google.errorprone.matchers.Matchers.toType;
+import static com.google.errorprone.matchers.method.MethodMatchers.staticMethod;
+
+import com.google.auto.service.AutoService;
+import com.google.auto.value.AutoValue;
+import com.google.common.collect.ImmutableList;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.annotations.Immutable;
+import com.google.errorprone.matchers.Matcher;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.ClassTree;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.source.tree.MethodTree;
+import com.sun.source.tree.Tree;
+import com.sun.source.util.TreeScanner;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.regex.Pattern;
+import org.jspecify.annotations.Nullable;
+
+/**
+ * An {@link Extractor} that describes how to extract data from classes that test a {@code
+ * BugChecker}.
+ */
+// XXX: Consider whether to omit or handle differently identification tests without `// BUG:
+// Diagnostic (contains|matches)` markers.
+// XXX: Handle other methods from `{BugCheckerRefactoring,Compilation}TestHelper`.
+@Immutable
+@AutoService(Extractor.class)
+@SuppressWarnings("rawtypes" /* See https://github.com/google/auto/issues/870. */)
+public final class BugPatternTestExtractor implements Extractor<BugPatternTestExtractor.TestCases> {
+  private static final Pattern TEST_CLASS_NAME_PATTERN = Pattern.compile("(.*)Test");
+  private static final Matcher<Tree> JUNIT_TEST_METHOD =
+      toType(MethodTree.class, hasAnnotation("org.junit.jupiter.api.Test"));
+
+  /** Instantiates a new {@link BugPatternTestExtractor} instance. */
+  public BugPatternTestExtractor() {}
+
+  @Override
+  public String identifier() {
+    return "bugpattern-test";
+  }
+
+  @Override
+  public Optional<BugPatternTestExtractor.TestCases> tryExtract(
+      ClassTree tree, VisitorState state) {
+    return getClassUnderTest(tree)
+        .map(
+            bugPatternName -> {
+              CollectBugPatternTests scanner = new CollectBugPatternTests();
+
+              for (Tree m : tree.getMembers()) {
+                if (JUNIT_TEST_METHOD.matches(m, state)) {
+                  scanner.scan(m, state);
+                }
+              }
+
+              return scanner.getTests().isEmpty()
+                  ? null
+                  : new AutoValue_BugPatternTestExtractor_TestCases(
+                      bugPatternName, scanner.getTests());
+            });
```

Comment:
```suggestion
        .flatMap(
            bugPatternName -> {
              CollectBugPatternTests scanner = new CollectBugPatternTests();

              for (Tree m : tree.getMembers()) {
                if (JUNIT_TEST_METHOD.matches(m, state)) {
                  scanner.scan(m, state);
                }
              }
              
              return scanner.getTests().map(tests -> 
                new AutoValue_BugPatterTestExtractorTestCases(
                  bugPatternName, tests));
            });
```

Reply:
As mentioned offline, my suggestion sorta reduces readability! So I prefer the original I think :smile: